### ISSUE TITLE
CSV import doesn't work

### ISSUE THREAD

**Issue by Jibec** [WeblateOrg/weblate#1547]
### Steps to reproduce
1. Open a component
2. Download the strings as CSV
3. Edit the CSV content
4. Upload the CSV

### Actual behaviour
I changed a few strings (mostly apostrophes and one word) and it says : 38 chaînes importées depuis les fichiers (ignorées : 0, non trouvées 0, mises à jour 0).

### Expected behaviour
It should have a a few ingored, no "not found", and a few updated.
Here is the edited CSV file:
https://github.com/Jibec/Weblate-CSV-Demo/commit/b65dca802815b656876d1052be60265f107d0516

### Server configuration
Weblate 2.12 and 2.15 have both the same issue (confirmed on mailing list).

thanks a lot for your help

**Comment by tchize** [comment]
Hi,

did some fiddle with source code by adding lots of logs to nail down problem. Here is my understanding so far (note I'm not a python developer so i may be missing some clues)

using  CSV extract below, with a java properties UTF-8 based project. Just one item in CSV to make fiddling easier :p

```
"location","source","target","id","fuzzy","context","translator_comments","developer_comments"
"","Any cargo type","N'importe quel type de Cargo","","False","anyCargoType","",""
```

After importing CSV, in the python csv structure, the id of csv entries become u'[context]\x04[source string]' (as seen from keys in id_index dict in the Translate object) and the importer never finds it. If you edit your CSV and copy the context to id column:

```
"location","source","target","id","fuzzy","context","translator_comments","developer_comments"
"","Any cargo type","N'importe quel type de Cargo","anyCargoType","False","anyCargoType","",""
```

 it goes one step further an recognize that string as worth a check.

> Processed 79 strings from the uploaded files (skipped: 0, not found: 1, updated: 0)
Also not that my csv file had only 1 translation, the number of String returned is actually the number of entries in the translation database requiring translation, so something might be fishy there (inversion of upload content vs database content?)

Now the further problem appear in translation.py -> merge_translations
in that method, self.unit_set has the value of None. I couldn't find in code where it's supposed to be something else, but as a result, call to self.unit_set.get_unit(unit2) fail to find translation entry corresponding to item.

This is all i could gather as information with my knowledge. If anyone has other idea what to look at? :)

**Comment by licaon-kter** [comment]
I had the same issue on https://hosted.weblate.org/projects

**Comment by nijel** [comment]
Apparently constructing ID is what translate-toolkit does: https://github.com/translate/translate/blob/master/translate/storage/csvl10n.py#L78

This is really not expected from Weblate side, so it might be cause of some problems.

**Comment by nijel** [comment]
I've addressed this bit in 5ea2805e4, but I don't have time to do deeper testing right now.

**Comment by licaon-kter** [comment]
Any idea when will this be deployed on https://hosted.weblate.org/ ?

**Comment by nijel** [comment]
Probably later today as I will be pushing some other changes as well.

**Comment by Jibec** [comment]
hi, thanks for this commit, it really makes the situation better!

I did a test with the project "Dino" on hosted.weblate.org, and it now says: 82 chaînes importées depuis les fichiers (ignorées : 0, non trouvées 10, mises à jour 72).
So, 10 strings not found is weird.
I'm also surprised withe 72 update, I hope it doesn't say it created records in database ;)

The situation is better, but still weird. In the upload form, I used the default settings.

Here is the file:
[dino-translations-fr.txt](https://github.com/WeblateOrg/weblate/files/1150912/dino-translations-fr.txt)

**Comment by licaon-kter** [comment]
Still not working for me.

Here is my file: https://gist.github.com/licaon-kter/2708931ebc989087648048e4b4f3a945

**Comment by Jibec** [comment]
Le 5 août 2017 13:32:58 UTC+03:00, "Michal Čihař" <notifications@github.com> a écrit :
>Closed #1547 via 4cdf14489dd66ae4f5c3d6cd3f8babbe2a72e4d2.

♡ thanks! Is this part of 2.16? Can I already try it on hosted?
-- 
Jean-Baptiste Holcroft

**Comment by nijel** [comment]
Thank you for your report, the issue you have reported has just been fixed. 

* In case you see problem with the fix, please comment on this issue. 
* In case you see similar problem, please open separate issue.
* If you are happy with the outcome, consider supporting Weblate by [donating](https://weblate.org/donate/).

**Comment by licaon-kter** [comment]
Still `Processed 372 strings from the uploaded files (skipped: 0, not found: 0, updated: 0).` :(
(or it was not yet deployed in spite of the previous message)

**Comment by Jibec** [comment]
Le 6 août 2017 00:32:53 UTC+03:00, licaon-kter <notifications@github.com> a écrit :
>Still `Processed 372 strings from the uploaded files (skipped: 0, not
>found: 0, updated: 0).` :(
>(or it was not yet deployed in spite of the previous message)

Hello, sorry but some basic information are needed: which project/language have you tried? When? Where is the file? How many strings have you changed and/or translated?
-- 
Jean-Baptiste Holcroft

**Comment by licaon-kter** [comment]
https://hosted.weblate.org/projects/f-droid/f-droid/ro/

The file is in my post above from 17.07.

I don't know how many, I'd have to do a `diff` for that.

Tested before posting here of course.

**Comment by nijel** [comment]
It was not yet deployed, I somehow didn't get to it. Anyway now it's there....

**Comment by licaon-kter** [comment]
Looks ok now. Thanks

**Comment by safi-manar** [comment]
Hello @nijel ,
This issue was actually a critical bug to my org, so we took it upon ourselves to invest a week to fix it. I was actually minutes away from submitting a pull request on Saturday when I noticed you had submitted the fix! :)

Anyway, when I looked at your changes, they seemed different then the ones I had put on my fork. I haven't yet tested yours, but when my preliminary tests on my fork worked. Let me know if you have the time and wanted to check out how we fixed it on our fork.
